src/app/clusters/wifi-network-management-server/wifi-network-management-server.cpp

```diff
@@ -0,0 +1,142 @@
+
+#include <functional>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::WiFiNetworkManagement::Attributes;
+using namespace std::placeholders;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace WiFiNetworkManagement {
+
+Global<Server> gServerInstance;
+
+Server & Server::Instance()
+{
+    return gServerInstance.get();
+}
+
+Server::Server() : AttributeAccessInterface(NullOptional, Id), CommandHandlerInterface(NullOptional, Id) {}
+
+CHIP_ERROR Server::Init(EndpointId endpoint)
+{
+    VerifyOrReturnError(endpoint != kInvalidEndpointId, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mEndpointId == kInvalidEndpointId, CHIP_ERROR_INCORRECT_STATE);
+
+    mEndpointId = endpoint;
+    VerifyOrReturnError(registerAttributeAccessOverride(this), CHIP_ERROR_INTERNAL);
+    ReturnErrorOnFailure(InteractionModelEngine::GetInstance()->RegisterCommandHandler(this));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Server::ClearNetworkCredentials()
+{
+    VerifyOrReturnError(!SsidSpan().empty() || !PassphraseSpan().empty(), CHIP_NO_ERROR);
+
+    mSsidLen = 0;
+    mPassphrase.SetLength(0);
+    MatterReportingAttributeChangeCallback(mEndpointId, Id, Ssid::Id);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Server::SetNetworkCredentials(ByteSpan ssid, ByteSpan passphrase)
+{
+    VerifyOrReturnError(1 <= ssid.size() && ssid.size() <= sizeof(mSsid), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(1 <= passphrase.size() && passphrase.size() <= mPassphrase.Capacity(), CHIP_ERROR_INVALID_ARGUMENT);
+
+    VerifyOrReturnError(!SsidSpan().data_equal(ssid) || !PassphraseSpan().data_equal(passphrase), CHIP_NO_ERROR);
+
+    memcpy(mSsid, ssid.data(), ssid.size());
+    mSsidLen = static_cast<decltype(mSsidLen)>(ssid.size());
+
+    ReturnErrorOnFailure(mPassphrase.SetLength(passphrase.size()));
+    memcpy(mPassphrase.Bytes(), passphrase.data(), passphrase.size());
+
+    MatterReportingAttributeChangeCallback(mEndpointId, Id, Ssid::Id); // report SSID change even if only passphrase changed
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Server::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case Ssid::Id: {
+        auto ssid = SsidSpan();
+        return (ssid.empty()) ? aEncoder.EncodeNull() : aEncoder.Encode(ssid);
+    }
+    }
+    return CHIP_NO_ERROR;
+}
+
+void Server::InvokeCommand(HandlerContext & ctx)
+{
+    switch (ctx.mRequestPath.mCommandId)
+    {
+    case Commands::NetworkPassphraseRequest::Id:
+        HandleCommand<Commands::NetworkPassphraseRequest::DecodableType>(
+            ctx, std::bind(&Server::HandleNetworkPassphraseRequest, this, _1, _2));
+        return;
+    }
+}
+
+void Server::HandleNetworkPassphraseRequest(HandlerContext & ctx, const Commands::NetworkPassphraseRequest::DecodableType & req)
+{
+    if (mPassphrase.Length() > 0)
```

Comment:
I don't see this in the spe.  The spec says you just return whatever the passphrase is.  There is no mention of InvalidInState in the spec that I can see...

Am I looking at the wrong spec?  I'm looking at what's on tip of `master` right now.

Reply:
The spec currently allows the SSID to be null (e.g. when the network is not set up yet). In that case we also don't have a passphrase to return, so I think we need to amend the spec to define what the status code should be in that case. I've filed https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/9234 to track this.

Reply:
Also to clarify further, I'm using `mSsidLen == 0` and `mPassphrase.Length() == 0` to represent the state where no network is configured yet. Possibly this check should be for `mSsidLen == 0` if we want to allow a passphrase length of 0 as a valid state, however according to the interpretation in the Network Commissioning Cluster ("Valid Credentials length are:" ...) this implies an unsecured / open network, which I don't think we want to allow here.

I will add this to the spec issue... We should really only be allowing WPA2-Personal or better credentials for the primary Wi-Fi Network for the home.

Reply:
Actually the spec says "WPA-Personal" so the credential should always be 8..63 bytes or 64 hex characters (but this could be made more explicit in the spec). `SetNetworkCredentials` should explicitly check for this.

Reply:
Added a `HaveNetworkCredentials()` helper to make the code clearer, as well as the described validation in `SetNetworkCredentials()`

Reply:
Looks great, thanks!